### PR TITLE
[WIP] fixes #27022 - settings via env variables

### PR DESCRIPTION
**Pull request by timogoebel**
This is still a very early version of the patch.

@ekohl & @ohadlevy: I'm using the definition of the default settings to guess the type of a setting. Is it safe to assume that every setting has a default setting defined? Or would we need to extend the plugin dsl to define settings and their type?

### CODE REVIEW

File: lib/proxy/settings.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ module Proxy::Settings
   def self.initialize_global_settings(settings_path = nil, argv = ARGV)
     global = ::Proxy::Settings::Global.new(YAML.load(File.read(settings_path || SETTINGS_PATH)))
     global.apply_argv(argv)
+    global.apply_env(ENV)
```
Comment (ekohl) [inline lib/proxy/settings.rb:13]: Would it make sense to make this a parameter as well, just like argv so it's easier to test?

File: lib/proxy/settings/global.rb [hunk 2]
```diff
@@ -53,5 +53,51 @@ def apply_argv(args)
       self.daemon = true if args.include?('--daemonize')
       self.daemon = false if args.include?('--no-daemonize')
     end
+
+    def apply_env(env)
+      DEFAULT_SETTINGS.each do |setting_key, setting_value|
+        env_key = "PROXY_#{setting_key.to_s.upcase}"
```
Comment (ekohl) [inline lib/proxy/settings/global.rb]: It may be longer, but maybe `FOREMAN_PROXY_` or `SMART_PROXY_`? Proxy can be ambiguous and I wonder if it can interfere with something.

**Review by ekohl — COMMENTED** [review]
Plugins have their own settings. Any plan to incorporate that? I was thinking about `FOREMAN_PROXY_<plugin>_<setting>`. I like the idea of a DSL for it, but I can tell you that there are settings without defaults. IMHO a `setting` DSL would be good. Perhaps something like:
```ruby
setting :puppet_url, cast: :url, required: true, expose: true
setting :puppet_key, cast: :path, required: true
setting :puppet_cert, cast: :path, required: true
setting :puppet_ca, cast: :path, required: true, default: '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
```

File: lib/proxy/settings.rb (hunk 1 above)
Comment (timogoebel) [reply lib/proxy/settings.rb:13]: It already is a parameter, for exactly that reason. ;-)

**Comment by timogoebel** [comment]
> Plugins have their own settings. Any plan to incorporate that?

Yes, definitely.

> IMHO a `setting` DSL would be good

This would require changes across all plugins/modules. I'd love to have a solution that works without any changes to the plugin DSL. Ideally, all settings have a default. Or can have one. I'll work on this more and we'll see.

**Comment by timogoebel** [comment]
Added settings for plugins.

TODOs:
- [x] add tests
- [ ] add missing default settings for all modules

**Comment by timogoebel** [comment]
Basically, we have two options to choose from:

1. Require all settings to have `default_settings` set. If the default value is nil, the setting is treated as optional. If settings are defined in a config file but aren't specified as default settings, log a warning. These settings cannot be overridden by an env variable.

2. Add a new settings dsl and add a compatibility layer that adds settings definition based on default settings etc. Eventually, deprecate the current settings api.

Basically, 1 is already implemented and the least intrusive option. 2 would probably be the cleanest solution, but is a lot more effort for little gain.

@ekohl: What do you think?

File: lib/proxy/plugin_validators.rb [hunk 3]
```diff
@@ -8,7 +8,7 @@ def initialize(plugin, setting_name, params, predicate)
     end
 
     def required_setting?
-      @plugin.plugin_default_settings.has_key?(@setting_name)
+      @plugin.plugin_default_settings.has_key?(@setting_name) && !@plugin.plugin_default_settings[@setting_name].nil?
```
Comment (iNecas) [inline lib/proxy/plugin_validators.rb:11]: Why do we need this change?

Comment (timogoebel) [reply lib/proxy/plugin_validators.rb:11]: Prior to this commit, not specifying a setting at all means that the setting is optional. For the setting to show up as an env variable, we need to explicitly define all settings. So setting the default to `nil` now means the setting is optional. Makes sense?

File: lib/proxy/settings_from_env.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      :list
+    when TrueClass, FalseClass
+      :boolean
+    when Float
+      :float
+    when Integer
+      :integer
+    when Symbol
+      :symbol
+    else
```
Comment (iNecas) [inline lib/proxy/settings_from_env.rb:14]: Isn't `Hash/:dict` missing here?

**Comment by iNecas** [comment]
Waiting for the all the changes before testing this fully. One small comment code-wise

**Comment by ares** [comment]
@timogoebel we started to also triage proxy PRs and wanted to check, if you see yourself working on this in foreseeable future. It's still labeled as WIP and there seems to be no update since June. Thanks for letting us know.

**Comment by timogoebel** [comment]
I'll close this for the time being.
